x-pack/filebeat/input/http_endpoint/handler.go

```diff
@@ -119,29 +146,88 @@ func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 				break
 			} else if !errors.Is(err, errNotCRC) {
 				h.metrics.apiErrors.Add(1)
-				h.sendAPIErrorResponse(w, r, h.log, http.StatusBadRequest, err)
+				h.sendAPIErrorResponse(txID, w, r, h.log, http.StatusBadRequest, err)
 				return
 			}
 		}
 
-		if err = h.publishEvent(obj, headers); err != nil {
+		acker.Add()
+		if err = h.publishEvent(obj, headers, acker); err != nil {
 			h.metrics.apiErrors.Add(1)
-			h.sendAPIErrorResponse(w, r, h.log, http.StatusInternalServerError, err)
+			h.sendAPIErrorResponse(txID, w, r, h.log, http.StatusInternalServerError, err)
 			return
 		}
 		h.metrics.eventsPublished.Add(1)
 		respCode, respBody = h.responseCode, h.responseBody
 	}
 
-	h.sendResponse(w, respCode, respBody)
-	if h.reqLogger != nil {
-		h.logRequest(r, respCode, nil)
+	acker.Ready()
+	if acked == nil {
+		h.sendResponse(w, respCode, respBody)
+	} else {
+		select {
+		case <-acked:
+			h.log.Debugw("request acked", "tx_id", txID)
+			if !timeout.Stop() {
+				<-timeout.C
+			}
+			h.sendResponse(w, respCode, respBody)
+		case <-timeout.C:
+			h.log.Debugw("request timed out", "tx_id", txID)
+			h.sendAPIErrorResponse(txID, w, r, h.log, http.StatusGatewayTimeout, errTookTooLong)
+		case <-h.ctx.Done():
+			h.log.Debugw("request context cancelled", "tx_id", txID)
+			h.sendAPIErrorResponse(txID, w, r, h.log, http.StatusGatewayTimeout, h.ctx.Err())
+		}
+		if h.reqLogger != nil {
+			h.logRequest(txID, r, respCode, nil)
+		}
 	}
 	h.metrics.batchProcessingTime.Update(time.Since(start).Nanoseconds())
 	h.metrics.batchesPublished.Add(1)
 }
 
-func (h *handler) sendAPIErrorResponse(w http.ResponseWriter, r *http.Request, log *logp.Logger, status int, apiError error) {
+var errTookTooLong = errors.New("could not publish event within timeout")
+
+func getTimeoutWait(u *url.URL, log *logp.Logger) (time.Duration, error) {
+	q := u.Query()
+	switch len(q) {
+	case 0:
+		return 0, nil
+	case 1:
+		if _, ok := q["wait_for_completion_timeout"]; !ok {
+			var k string
+			for k = range q {
+				break
+			}
+			return 0, fmt.Errorf("unexpected URL query: %s", k)
+		}
+	default:
+		delete(q, "wait_for_completion_timeout")
+		keys := make([]string, 0, len(q))
+		for k := range q {
+			keys = append(keys, k)
+		}
+		sort.Strings(keys)
+		return 0, fmt.Errorf("unexpected URL query: %s", strings.Join(keys, ", "))
+	}
+	p := q.Get("wait_for_completion_timeout")
+	if p == "" {
+		// This will never happen; it is already handled in the check switch above.
+		return 0, nil
+	}
+	log.Debugw("wait_for_completion_timeout parameter", "value", p)
```

Comment:
Same thing here, every incoming request can potentially generate two log lines, in a high throughput environment, that might flood the logs.

Reply:
My debugging experience with inputs has nearly always been that I'm lacking information due to non-logging of things like this rather than information flood. Note that it is DEBUG; ideally we'd have DEBUG-1, but this serves.